lib/membrane/timestamp_queue.ex

```diff
@@ -0,0 +1,389 @@
+    |> push_item(pad_ref, :end_of_stream)
+    |> put_in([:pad_queues, pad_ref, :end_of_stream?], true)
+  end
+
+  defp push_item(%__MODULE__{} = timestamp_queue, pad_ref, item) do
+    timestamp_queue
+    |> maybe_handle_item_from_new_pad(item, pad_ref)
+    |> update_in(
+      [:pad_queues, pad_ref, :qex],
+      &Qex.push(&1, item)
+    )
+  end
+
+  defp maybe_handle_item_from_new_pad(
+         %__MODULE__{pad_queues: pad_queues} = timestamp_queue,
+         _item,
+         pad_ref
+       )
+       when is_map_key(pad_queues, pad_ref) do
+    timestamp_queue
+  end
+
+  defp maybe_handle_item_from_new_pad(%__MODULE__{} = timestamp_queue, first_item, pad_ref) do
+    priority =
+      case first_item do
+        {:buffer, _buffer} -> -timestamp_queue.current_queue_time
+        _other -> :infinity
+      end
+
+    timestamp_queue
+    |> put_in([:pad_queues, pad_ref], new_pad_queue())
+    |> Map.update!(:pads_heap, &Heap.push(&1, {priority, pad_ref}))
+  end
+
+  defp new_pad_queue() do
+    %{
+      timestamp_offset: nil,
+      qex: Qex.new(),
+      buffers_size: 0,
+      buffers_number: 0,
+      update_heap_on_buffer?: true,
+      paused_demand?: false,
+      end_of_stream?: false,
+      use_pts?: nil,
+      max_timestamp_on_qex: nil,
+      recently_returned_timestamp: nil
+    }
+  end
+
+  defp actions_after_pushing_buffer(pad_queue, pad_ref, pause_demand_boundary) do
+    if not pad_queue.paused_demand? and pad_queue.buffers_size >= pause_demand_boundary do
+      {[pause_auto_demand: pad_ref], %{pad_queue | paused_demand?: true}}
+    else
+      {[], pad_queue}
+    end
+  end
+
+  defp buffer_time(%Buffer{dts: dts}, %{use_pts?: false, timestamp_offset: timestamp_offset}),
+    do: dts + timestamp_offset
+
+  defp buffer_time(%Buffer{pts: pts}, %{use_pts?: true, timestamp_offset: timestamp_offset}),
+    do: pts + timestamp_offset
+
+  @type item ::
+          {:stream_format, StreamFormat.t()}
+          | {:buffer, Buffer.t()}
+          | {:event, Event.t()}
+          | :end_of_stream
+
+  @type popped_value :: {Pad.ref(), item()}
+
+  @doc """
+  Pops as many buffers from the queue, as it is possible.
+
+  Returns suggested actions list, list of popped buffers and the updated queue.
+
+  If the amount of buffers associated with any pad in the queue falls below the
+  `pause_demand_boundary`, the suggested actions list contains `t:Action.resume_auto_demand()`
+  actions, otherwise it is an empty list.
+
+  If the queue cannot return any buffer, empty list is returned. Note, that queue doesn't have to be
+  empty to be unable to return a buffer - sometimes queue must keep up to 1 buffer for each pad,
+  to be able to work correctly.
+
+  To be able to maintain proper order of buffers from varius pads, queue won't return next buffer,
+  if the next buffer that should be returned:
+   - is the only buffer from a certain pad
+   - and that certain pad has not received the end of stream yet
```

Comment:
Here's my try, correct me if I'm wrong:

```suggestion
  Pops items from the queue while they are available.
  
  An item that is not a buffer is always considered available. A buffer is
  available when the following conditions are met:
  - There is another buffer or `end_of_stream` enqueued on the pad
  - On each other pad there is either `end_of_stream` or a buffer with a lower timestamp.

  The returned value is a suggested actions list, a list of popped buffers and the updated queue.

  If the amount of buffers associated with any pad in the queue falls below the
  `pause_demand_boundary`, the suggested actions list contains `t:Action.resume_auto_demand()`
  actions, otherwise it is an empty list.
```